### PR TITLE
Refactor how child chunks are stored.

### PR DESCRIPTION
**Pull request by munificent**
Given a piece of code like:

    main() {
      a();
      b();
    }

There are top level chunks for `main() }` and `}`, and the chunks for
`a()` and `b()` are children.

Previously, the code stored those child chunks in the preceding parent
chunk (so `main() {` here). But it's the subsequent chunk (`}`) that
determines whether the block contents are actually split, so that
doesn't make a lot of sense and leads to weird `+ 1` and `- 1` when
working with nested chunks.

This refactors the code so that child chunks are stored on the same
chunk that determines whether or not they split. This means that chunks
are now written in a post-order traversal: a block chunk's children come
before its own text.

Since we now know that a chunk will have children at the moment that
it's created, removed the old ChunkBlock class and replaced it with a
BlockChunk subclass of Chunk.

Also added a bunch of tests around trailing whitespace in blocks. When I
was testing this change on a corpus of code, I thought it inadvertently
changed some behavior, but it turns out that it was the previous
refactoring (which did deliberately change block formatting) and not
this one. These tests help pin that behavior down.

This commit here has zero changes on the formatter's visible behavior.

### CODE REVIEW

**Review by natebosch — APPROVED** [review]

File: lib/src/chunk.dart [hunk 1]
```diff
@@ -189,29 +170,9 @@ class Chunk extends Selection {
...
-
-    // There may be no rule if the block occurs inside a string interpolation.
-    // In that case, it's not clear if anything will look particularly nice, but
-    // expression nesting is probably marginally better.
-    if (rule == Rule.dummy) return true;
-
-    return rule.isSplit(getValue(rule), argument);
-  }
+  /// Returns `true` if this chunk is a block whose children should be
+  /// expression indented given a set of rule values provided by [getValue].
```
[the remark this answers is not in the document]
Comment (munificent) [reply lib/src/chunk.dart:174]: The documentation is a little spread across the codebase, but the basic idea is that a Rule object represents a choice that the LineSplitter must make about how line breaks should be inserted. Each Rule can have one of a number of values determined by that particular Rule. Each value determines which chunk using that Rule get split. In simple rules, the value 0 means "no chunks split" and 1 means "all chunks split". For things like ArgumentRules, there may be more values to represent other cases like "split before first argument", "split before all arguments", etc.

This callback returns the value for the given rule. Added a brief comment explaining.
